gradient/commands/datasets.py

```diff
@@ -558,144 +558,146 @@ def update_status():
-                                'params': {
-                                    'Key': key,
-                                    'UploadId': mpu_data['UploadId'],
-                                    'MultipartUpload': {'Parts': parts}
-                                }
-                            }]
-                        }
-                    )
-
-                self.validate_s3_response(r)
-            except requests.exceptions.ConnectionError as e:
-                return self.report_connection_error(e)
-            except Exception as e:
-                return e
+        try:
+            if size <= 0:
+                headers.update({'Content-Size': '0'})
+                r = session.put(url, data='', headers=headers, timeout=5)
+            # for files under 15MB
+            elif size <= (MULTIPART_CHUNK_SIZE):
+                with open(path, 'rb') as f:
+                    r = session.put(
+                        url, data=f, headers=headers, timeout=PUT_TIMEOUT)
+            # # for chonky files, use a multipart upload
+            else:
+                # Chunks need to be at least 5MB or AWS throws an
+                # EntityTooSmall error; we'll arbitrarily choose a
+                # 15MB chunksize
+                #
+                # Note also that AWS limits the max number of chunks
+                # in a multipart upload to 10000, so this setting
+                # currently enforces a hard limit on 150GB per file.
+                #
+                # We can dynamically assign a larger part size if needed,
+                # but for the majority of use cases we should be fine
+                # as-is
+                part_minsize = MULTIPART_CHUNK_SIZE
+                dataset_id, _, version = dataset_version_id.partition(":")
+                mpu_url = f'/datasets/{dataset_id}/versions/{version}/s3/preSignedUrls'
+                api_client = http_client.API(
+                    api_url=config.CONFIG_HOST,
+                    api_key=self.api_key,
+                    ps_client_name=CLI_PS_CLIENT_NAME
+                )
+
+                mpu_create_res = api_client.post(
+                    url=mpu_url,
+                    json={
+                        'datasetId': dataset_id,
+                        'version': version,
+                        'calls': [{
+                            'method': 'createMultipartUpload',
+                            'params': {'Key': key}
+                        }]
+                    }
+                )
+
+                mpu_data = mpu_create_res.json()[0]['url']
+
+                parts = []
+                with open(path, 'rb') as f:
+                    # we +2 the number of parts since we're doing floor
+                    # division, which will cut off any trailing part
+                    # less than the part_minsize, AND we want to 1-index
+                    # our range to match what AWS expects for part
+                    # numbers
```

Comment:
Is this true? Shouldn't you use `ceil`? I have a `gradient version create` and `gradient files put` where it crashes when there is a file that evenly divides the 15mb chunk size (75mb). I suspect that it's trying to read an extra chunk that doesn't exist. Does that make sense? I could be misreading things.

Also I think you mean to say that you `1+index` because you start counting from 1 and you need to correct for `range`. Not entirely sure. Could you check this out?

Reply:
@jared-paperspace

Reply:
Looks like this was a known bug in the PR but intentionally left in. 🥇 

https://github.com/Paperspace/gradient-cli/pull/384#discussion_r846314866